sup/gpu_uuid.cpp

```diff
@@ -0,0 +1,336 @@
+    int ngpus = 0;
+    auto cuda_status = cudaGetDeviceCount(&ngpus);
+    if (cuda_status!=cudaSuccess) throw make_runtime_error(cuda_status);
+
+    // Attempt to initialize nvml
+    auto nvml_status = nvmlInit();
+    if (nvml_status!=NVML_SUCCESS) throw make_runtime_error(nvml_status);
+    auto nvml_guard = on_scope_exit(nvmlShutdown);
+
+    // store the uuids
+    std::vector<uuid> uuids;
+
+    // find the number of available GPUs
+    unsigned count = -1;
+    nvml_status = nvmlDeviceGetCount(&count);
+    if (nvml_status!=NVML_SUCCESS) throw make_runtime_error(nvml_status);
+
+    // Indexes of GPUs available on this rank
+    std::vector<int> device_ids;
+
+    // Test if the environment variable CUDA_VISIBLE_DEVICES has been set.
+    const char* visible_device_env = std::getenv("CUDA_VISIBLE_DEVICES");
+    // If set, attempt to parse the device ids from it.
+    if (visible_device_env) {
+        // Parse the gpu ids from the environment variable
+        device_ids = parse_visible_devices(visible_device_env, count);
+        if ((unsigned)ngpus != device_ids.size()) {
+            // Mismatch between device count detected by cuda runtime
+            // and that set in environment variable.
+            throw std::runtime_error(
+                "Mismatch between the number of devices in CUDA_VISIBLE_DEVICES"
+                " and the number of devices detected by cudaGetDeviceCount.");
+        }
+    }
+    // Not set, so all devices must be available.
+    else {
+        device_ids.resize(count);
+        std::iota(device_ids.begin(), device_ids.end(), 0);
+    }
+
+    // For each device id, query NVML for the device's uuid.
+    for (int i: device_ids) {
+        char buffer[41];
+        // get handle of gpu with index i
+        nvmlDevice_t handle;
+        nvml_status = nvmlDeviceGetHandleByIndex(i, &handle);
+        if (nvml_status!=NVML_SUCCESS) throw make_runtime_error(nvml_status);
+
+        // get uuid as a string with format GPU-xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+        nvml_status = nvmlDeviceGetUUID(handle, buffer, sizeof(buffer));
+        if (nvml_status!=NVML_SUCCESS) throw make_runtime_error(nvml_status);
+
+        uuids.push_back(string_to_uuid(buffer));
+    }
+
+    return uuids;
+}
+#endif
+
+// Compare two sets of uuids
+//   1: both sets are identical
+//  -1: some common elements
+//   0: no common elements
+// Each set is described by a pair of iterators.
+template <typename R>
+int compare_gpu_groups(R l, R r) {
+    auto range_size = [] (auto& rng) { return std::distance(rng.first, rng.second);};
+    if (range_size(l)<range_size(r)) {
+        std::swap(l, r);
+    }
+
+    unsigned count = 0;
+    for (auto it=l.first; it!=l.second; ++it) {
+        if (std::find(r.first, r.second, *it)!=r.second) ++count;
+    }
+
+    // test for complete match
+    if (count==range_size(l) && count==range_size(r)) return 1;
+    // test for partial match
+    if (count) return -1;
+    return 0;
+}
+
```

Comment:
Because this is a simpler algorithm and `uuids_per_rank` will always be a small number.
For such utility code, I don't think that time spent making a faster algorithm that is harder to understand is a good investment.